sqlsmith: non-deterministic query generation

pkg/ccl/testccl/sqlccl/sqlccl_test.TestExplainGist [failed](https://tanzanite.cluster.engflow.com/invocations/default/8fbe1ff7-7f18-4b42-860d-6d43bd03d00a?testReportRun=8&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvc3FsY2NsOnNxbGNjbF90ZXN0) on release-24.3 @ [047a8e99eb7a35e3f13ecf6b113f1965f206e3b9](https://github.com/cockroachdb/cockroach/commits/047a8e99eb7a35e3f13ecf6b113f1965f206e3b9):


```
        WITH
        	with_162 (col_1215)
        		AS (
        			SELECT
        				*
        			FROM
        				(VALUES ('infinity':::DATE), ('5874897-12-31':::DATE), (NULL), ('1986-12-30':::DATE))
        					AS tab_824 (col_1215)
        		),
        	with_163 (col_1216) AS (SELECT * FROM (VALUES (NULL)) AS tab_825 (col_1216))
        SELECT
        	0.916923857798615:::FLOAT8 AS col_1217,
        	'1993-01-08 03:22:41.000793+00':::TIMESTAMPTZ AS col_1218,
        	CASE
        	WHEN false
        	THEN ('-77 years -11 mons -826 days -19:25:27.108467':::INTERVAL::INTERVAL * (-9223372036854775807):::INT8::INT8)::INTERVAL
        	ELSE '35 years 5 mons 637 days 22:29:35.3281':::INTERVAL
        	END
        		AS col_1219,
        	false AS col_1220,
        	1:::DECIMAL AS col_1221,
        	'!{^n':::STRING AS col_1222,
        	'09710b14-086c-4225-861c-edd9267108e1':::UUID AS col_1223,
        	NULL AS col_1224,
        	'{"bar": 1.0271062249458585, "fVSFxi": {"Ntqq": 1.331374259127252, "foobar": false}, "foo": 0.11168124355945516, "foobar": null}':::JSONB
        		AS col_1225
        FROM
        	with_163 AS cte_ref_52, with_163 AS cte_ref_53
        LIMIT
        	88:::INT8;
        UPDATE
        	defaultdb.public.tab_606 AS tab_826
        SET
        	col0_6 = '09:39:02.093446+10:32':::TIMETZ, col0_4 = (-0.3896579339753643):::FLOAT8
        FROM
        	defaultdb.public.seed AS tab_827
        LIMIT
        	94:::INT8;
        
    explain_test.go:202: pq: internal error: lookup join with no lookup columns: SELECT
        	'00000000-0000-0000-0000-000000000000':::UUID AS col_1226, NULL AS col_1227
        FROM
        	defaultdb.public.tab_606@tab_606_col0_11_key AS tab_828, defaultdb.public.tab_606@tab_606_col0_10_idx AS tab_829
        WHERE
        	true
        ORDER BY
        	tab_829.col0_6 NULLS LAST
        LIMIT
        	28:::INT8
    --- FAIL: TestExplainGist/main (5.06s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=8</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #133343 pkg/ccl/testccl/sqlccl/sqlccl_test: TestExplainGist failed [C-test-failure O-robot T-sql-queries branch-master release-blocker]
- #132828 pkg/ccl/testccl/sqlccl/sqlccl_test: TestExplainGist failed [C-test-failure O-robot T-sql-queries branch-release-24.2 release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExplainGist.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-43603

--- a/pkg/sql/sem/builtins/pg_builtins.go
+++ b/pkg/sql/sem/builtins/pg_builtins.go
@@ -8,6 +8,7 @@ package builtins
 import (
 	"context"
 	"fmt"
+	"sort"
 	"strconv"
 	"strings"
 	"time"
@@ -176,8 +177,16 @@ func init() {
 			},
 		)
 	})
-	// Add casts between the same type.
-	for typOID, def := range castBuiltins {
+	// Add casts between the same type in deterministic order.
+	typOIDs := make([]oid.Oid, 0, len(castBuiltins))
+	for typOID := range castBuiltins {
+		typOIDs = append(typOIDs, typOID)
+	}
+	sort.Slice(typOIDs, func(i, j int) bool {
+		return typOIDs[i] < typOIDs[j]
+	})
+	for _, typOID := range typOIDs {
+		def := castBuiltins[typOID]
 		typ := types.OidToType[typOID]
 		if !shouldMakeFromCastBuiltin(typ) {
 			continue

--- a/pkg/sql/sem/cast/cast.go
+++ b/pkg/sql/sem/cast/cast.go
@@ -10,6 +10,8 @@
 package cast
 
 import (
+	"sort"
+
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/volatility"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/lib/pq/oid"
@@ -128,14 +130,31 @@ type Cast struct {
 }
 
 // ForEachCast calls fn for every valid cast from a source type to a target
-// type.
+// type. Iteration order is deterministic.
 func ForEachCast(
 	fn func(
 		src oid.Oid, tgt oid.Oid, castCtx Context, ctxOrigin ContextOrigin, v volatility.V,
 	),
 ) {
-	for src, tgts := range castMap {
-		for tgt, cast := range tgts {
+	srcOids := make([]oid.Oid, 0, len(castMap))
+	for src := range castMap {
+		srcOids = append(srcOids, src)
+	}
+	sort.Slice(srcOids, func(i, j int) bool {
+		return srcOids[i] < srcOids[j]
+	})
+	var tgtOids []oid.Oid
+	for _, src := range srcOids {
+		tgts := castMap[src]
+		tgtOids = tgtOids[:0]
+		for tgt := range tgts {
+			tgtOids = append(tgtOids, tgt)
+		}
+		sort.Slice(tgtOids, func(i, j int) bool {
+			return tgtOids[i] < tgtOids[j]
+		})
+		for _, tgt := range tgtOids {
+			cast := tgts[tgt]
 			fn(src, tgt, cast.MaxContext, cast.origin, cast.Volatility)
 		}
 	}

